File: src/sim/rng.ts
/**
 * PRNG déterministe (mulberry32). À seed égale et commandes égales, une
 * bataille rejoue toujours exactement de la même façon : indispensable pour
 * les tests, le débogage, et plus tard pour valider les tirages côté serveur.
 */
export type Rng = () => number;

export function createRng(seed: number): Rng {
  let a = seed >>> 0;
  return function rng(): number {
    a |= 0;
    a = (a + 0x6d2b79f5) | 0;
    let t = Math.imul(a ^ (a >>> 15), 1 | a);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

export function pickRandom<T>(rng: Rng, items: readonly T[]): T {
  if (items.length === 0) {
    throw new Error("pickRandom: la liste est vide");
  }
  const index = Math.floor(rng() * items.length);
  return items[Math.min(index, items.length - 1)];
}

File: core/Healer.Combat/Rng.cs
using System;
using System.Collections.Generic;

namespace Healer.Combat
{
    /// <summary>
    /// PRNG déterministe (mulberry32), identique à la version TypeScript : à seed égale et
    /// commandes égales, un combat rejoue exactement de la même façon. C'est le SEUL générateur
    /// aléatoire autorisé dans le cœur (System.Random est interdit, vérifié par test).
    /// </summary>
    public sealed class Rng
    {
        private uint _a;

        public Rng(uint seed)
        {
            _a = seed;
        }

        /// <summary>Nombre dans [0, 1).</summary>
        public double Next()
        {
            unchecked
            {
                _a += 0x6D2B79F5u;
                uint t = (_a ^ (_a >> 15)) * (1u | _a);
                t = (t + (t ^ (t >> 7)) * (61u | t)) ^ t;
                return (t ^ (t >> 14)) / 4294967296.0;
            }
        }

        public T PickRandom<T>(IReadOnlyList<T> items)
        {
            if (items.Count == 0) throw new InvalidOperationException("PickRandom : la liste est vide");
            int index = (int)Math.Floor(Next() * items.Count);
            return items[Math.Min(index, items.Count - 1)];
        }
    }
}
